controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -347,41 +324,117 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
+	default:
+		machinesToDelete := internal.FilterMachines(selectedForUpgrade, internal.WithoutDeletionTimestamp())
+		if len(machinesToDelete) != len(selectedForUpgrade) {
+			logger.Info("waiting for already deleted machines to finish deleting")
+			return ctrl.Result{RequeueAfter: DeleteRequeueAfter}, nil
+		}
+		machineToDelete := oldestMachine(machinesToDelete)
+		if machineToDelete == nil {
+			logger.Error(err, "failed to pick control plane Machine to delete")
+			return ctrl.Result{}, err
+		}
+
+		logger := logger.WithValues("machine", machineToDelete.Name)
+		if err := r.managementCluster.RemoveEtcdMemberForMachine(ctx, clusterKey(cluster), machineToDelete); err != nil {
+			logger.Error(err, "failed to remove etcd member for machine")
+			return ctrl.Result{}, err
+		}
+		// TODO: Update the kubeadm configmap
+		if err := r.Client.Delete(ctx, machineToDelete); err != nil && !apierrors.IsNotFound(err) {
+			logger.Error(err, "failed to delete machine")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *KubeadmControlPlaneReconciler) selectMachineForUpgrade(ctx context.Context, cluster *clusterv1.Cluster, requireUpgrade []*clusterv1.Machine) (*clusterv1.Machine, error) {
+	failureDomain := r.failureDomainForUpgrade(cluster, requireUpgrade)
+	oldest := oldestMachine(internal.FilterMachines(requireUpgrade, internal.InFailureDomain(failureDomain)))
+	if oldest == nil {
+		if oldest == nil {
+			return nil, errors.New("failed to pick control plane Machine to delete")
+		}
+	}
+
+	patchHelper, err := patch.NewHelper(oldest, r.Client)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to create patch helper for machine %s", oldest.Name)
+	}
+
+	oldest.Labels[controlplanev1.SelectedForUpgradeLabel] = ""
+
+	if err := patchHelper.Patch(ctx, oldest); err != nil {
+		return nil, errors.Wrapf(err, "failed to patch machine %s selected for upgrade", oldest.Name)
+	}
+
+	return oldest, nil
 }
 
-func (r *KubeadmControlPlaneReconciler) initializeControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane) (ctrl.Result, error) {
+func (r *KubeadmControlPlaneReconciler) initializeControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, logger logr.Logger) (ctrl.Result, error) {
 	bootstrapSpec := kcp.Spec.KubeadmConfigSpec.DeepCopy()
 	bootstrapSpec.JoinConfiguration = nil
 
 	if err := r.cloneConfigsAndGenerateMachine(ctx, cluster, kcp, bootstrapSpec); err != nil {
-		return ctrl.Result{}, errors.Wrapf(err, "failed to create control plane Machine for cluster %s/%s", cluster.Name, cluster.Namespace)
+		logger.Error(err, "failed to create initial control plane Machine")
+		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "FailedInitialization", "Failed to create initial control plane Machine for cluster %s/%s control plane: %v", cluster.Namespace, cluster.Name, err)
+		return ctrl.Result{}, err
```

Comment:
General question about how errors are being handled:

Since the error is both being logged and recorded in an event, could this return `ctrl.Result{Requeue:true}, nil` instead?

Reply:
That is how I had initially implemented the error handling for KubeadmControlPlane and @ncdc requested that errors be returned to controller-runtime.